clippy_lints/src/methods/option_take_on_temporary.rs

```diff
@@ -0,0 +1,41 @@
+use clippy_utils::diagnostics::span_lint_and_sugg;
+use clippy_utils::match_def_path;
+use clippy_utils::source::snippet_with_applicability;
+use clippy_utils::ty::is_type_diagnostic_item;
+use rustc_errors::Applicability;
+use rustc_hir::Expr;
+use rustc_lint::LateContext;
+use rustc_span::sym;
+
+use super::OPTION_TAKE_ON_TEMPORARY;
+
+pub(super) fn check<'tcx>(cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>, recv: &'tcx Expr<'_>) {
+    // Checks if expression type is equal to sym::Option and if the expr is not a syntactic place
+    if is_expr_option(cx, recv) && !recv.is_syntactic_place_expr() && has_expr_as_ref_path(cx, recv) {
```

Comment:
I think `is_syntactic_place_expr` is faster than getting the type, so switching the order here should improve perf a small bit.
```suggestion
    if !recv.is_syntactic_place_expr() && is_expr_option(cx, recv) && has_expr_as_ref_path(cx, recv) {
```